### PR TITLE
Fabric checkpointing 1/n: base implementation

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the base implementation for checkpoint saving and loading, with the new API. Details of how it is used are in the docs pages of this PR. 
Follow up PRs will add the implementation in the DeepSpeed and FSDP strategies.

Part of #14816

```py
# Saving
state = {
    "model1": model1,
    "model2": model2,
    "optimizer": optimizer,
    "iteration": iteration,
}
fabric.save("./checkpoints/version_2/checkpoint.ckpt", state)

# Loading (selective, in-place)
state = {"model1": model1}
fabric.load("./checkpoints/version_2/checkpoint.ckpt", state)

# Loading (all)
all_states = fabric.load("./checkpoints/version_2/checkpoint.ckpt")
```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @awaelchli @carmocca @justusschock

### CODE REVIEW

File: src/lightning_fabric/strategies/strategy.py [hunk 1]
```diff
@@ -247,23 +248,36 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
 
-    def load_optimizer_state_dict(
-        self, optimizers: Union[Optimizer, Iterable[Optimizer]], checkpoint: Mapping[str, Any]
-    ) -> None:
-        if not isinstance(optimizers, Iterable):
-            optimizers = [optimizers]
-        optimizer_states = checkpoint["optimizer_states"]
-        for optimizer, opt_state in zip(optimizers, optimizer_states):
-            optimizer.load_state_dict(opt_state)
-            _optimizer_to_device(optimizer, self.root_device)
```
Comment (awaelchli) [inline src/lightning_fabric/strategies/strategy.py:266]: this was unused code.

File: docs/source-pytorch/fabric/guide/checkpoint.rst [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Fabric and the underlying strategy will decide in which format your checkpoint gets saved.
+For example, ``strategy="ddp"`` saves a single file on rank 0, while ``strategy="fsdp"`` saves multiple files from all ranks.
+
+
+----
+
+
+*************************
+Restore from a checkpoint
+*************************
```
Comment (awaelchli) [inline docs/source-pytorch/fabric/guide/checkpoint.rst:48]: These are very basic docs.

Once we have the deepspeed and fsdp implementations, I will create a document linked here that explains best practices and efficiency around saving and loading large sharded models.

File: src/lightning_fabric/strategies/strategy.py [hunk 3]
```diff
@@ -216,21 +216,22 @@ def reduce_boolean_decision(self, decision: bool, all: bool = True) -> bool:
...
         Args:
-            checkpoint: dict containing model and trainer state
-            filepath: write-target file's path
-            storage_options: parameter for how to save to storage, passed to ``CheckpointIO`` plugin
+            path: A path to where the file(s) should be saved
+            state: A dictionary with contents to be saved. If the dict contains modules or optimizers, their
+                state-dict will be retrieved and converted automatically.
+            storage_options: Additional options for the ``CheckpointIO`` plugin
         """
+        state = self._convert_stateful_objects_in_state(state)
```
Comment (awaelchli) [inline src/lightning_fabric/strategies/strategy.py:229]: For now, I have given the strategy the responsibility to handle the objects in the state. This is to make the deepspeed implementation easier, since it needs access to everything.

File: src/lightning_fabric/strategies/strategy.py [hunk 4]
```diff
@@ -247,23 +248,36 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
-        torch.cuda.empty_cache()
-        return self.checkpoint_io.load_checkpoint(checkpoint_path)
+    def load_checkpoint(
+        self, path: _PATH, state: Optional[Dict[str, Union[Module, Optimizer, Any]]] = None
+    ) -> Dict[str, Any]:
+        """Load the contents from a checkpoint and restore the state of the given objects.
 
-    def load_module_state_dict(self, module: Module, checkpoint: Mapping[str, Any]) -> None:
-        # TODO(fabric): Integrate this into Lightning Fabric
-        module.load_state_dict(checkpoint["state_dict"])
```
Comment (awaelchli) [inline src/lightning_fabric/strategies/strategy.py:256]: this was also unused code

File: src/lightning_fabric/strategies/strategy.py (hunk 3 above)
Comment (justusschock) [reply src/lightning_fabric/strategies/strategy.py:228]: I think as this is mainly strategy-dependent we could also leave it that way.

**Review by justusschock — APPROVED** [review]

File: docs/source-pytorch/fabric/api/fabric_methods.rst [hunk 5]
```diff
@@ -127,25 +127,55 @@ This avoids excessive printing and logs when running on multiple devices/nodes.
...
+        "model1": model1,
+        "model2": model2,
+        "optimizer": optimizer,
+        "iteration": iteration,
+    }
+
+    # Instead of `torch.save(...)`
+    fabric.save("path/to/checkpoint.ckpt", state)
+
+You can pass in the model and optimizer objects directly, and Fabric will unwrap them and retrieve their *state-dict* automatically.
```
Comment (carmocca) [inline docs/source-pytorch/fabric/api/fabric_methods.rst:147]: Isn't this a "must" more than a "can"?

File: src/lightning_fabric/wrappers.py [hunk 6]
```diff
@@ -189,3 +190,17 @@ def _process_optimizer_zero_grad_kwargs(optimizer: Optimizer, kwargs: Dict[str,
...
+def _unwrap_objects(collection: T_collection) -> T_collection:
+    def _unwrap(
+        obj: Union[_FabricModule, _FabricOptimizer, _FabricDataLoader]
+    ) -> Union[nn.Module, Optimizer, DataLoader]:
+        if isinstance(obj, _FabricModule):
+            return obj._forward_module
+        if isinstance(obj, _FabricOptimizer):
+            return obj.optimizer
+        if isinstance(obj, _FabricDataLoader):
+            return obj._dataloader
```
Comment (carmocca) [inline src/lightning_fabric/wrappers.py:204]: Should we have a default fallback? or a raise ValueError

```suggestion
            return obj._dataloader
        return obj
```

File: src/lightning_fabric/wrappers.py [hunk 7]
```diff
@@ -189,3 +190,17 @@ def _process_optimizer_zero_grad_kwargs(optimizer: Optimizer, kwargs: Dict[str,
         # Some optimizers out there, for example DeepSpeedZeroOptimizer, use a different name than PyTorch
         kwargs["set_grads_to_None"] = kwargs.pop("set_to_none")
     return kwargs
+
+
+def _unwrap_objects(collection: T_collection) -> T_collection:
```
Comment (carmocca) [inline src/lightning_fabric/wrappers.py:195]: This type annotation isn't correct, as you aren't returning the same collection. I think mypy doesn't complain because typing gets lost through `apply_to_collection` and treats it as `Any`.

So maybe you should annotate it as `Any`, or leave it as it is.

File: src/lightning_fabric/strategies/strategy.py [hunk 8]
```diff
@@ -247,23 +247,36 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
+        Returns:
+            The remaining items that were not restored into the given state dictionary. If no state dictionary is
+            given, the full checkpoint will be returned.
+        """
+        torch.cuda.empty_cache()
+        checkpoint = self.checkpoint_io.load_checkpoint(path)
+        if not state:
+            return checkpoint
+
+        for name, obj in state.copy().items():
```
Comment (carmocca) [inline src/lightning_fabric/strategies/strategy.py:269]: Why this `copy()`?

File: src/lightning_fabric/strategies/strategy.py [hunk 9]
```diff
@@ -247,23 +247,36 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
+            if name not in checkpoint:
+                continue
+            elif isinstance(obj, Module):
+                # TODO(fabric): Make strict loading configurable
+                obj.load_state_dict(checkpoint.pop(name), strict=True)
+            elif isinstance(obj, Optimizer):
```
Comment (carmocca) [inline src/lightning_fabric/strategies/strategy.py:275]: Should we use structural typing here? (as in #16189)

We have the `_Stateful` type for this

File: src/lightning_fabric/strategies/strategy.py [hunk 10]
```diff
@@ -247,23 +247,36 @@ def get_optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
...
+                obj.load_state_dict(checkpoint.pop(name))
+            else:
+                state[name] = checkpoint.pop(name)
```
Comment (carmocca) [inline src/lightning_fabric/strategies/strategy.py:278]: Why do you `pop` instead of accessing it without removal?

File: src/lightning_fabric/wrappers.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning_fabric/wrappers.py:204]: Sure, can't hurt, but not strictly necessary because apply_to_collection won't ever call this function not matching the given types.

File: src/lightning_fabric/strategies/strategy.py (hunk 8 above)
Comment (awaelchli) [reply src/lightning_fabric/strategies/strategy.py:269]: Because I'm modifying the keys as I am iterating over the dict.  I believe it is valid use here, do you see a problem?

`state[name] = checkpoint.pop(name)`

File: src/lightning_fabric/wrappers.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning_fabric/wrappers.py]: I knew I would get roasted on this haha.
Changed to Any

File: src/lightning_fabric/strategies/strategy.py (hunk 9 above)
Comment (carmocca) [reply src/lightning_fabric/strategies/strategy.py:275]: Yeah let's do it, better than nothing

```python
if isinstance(obj, _Stateful):
    if isinstance(obj, Module):
        obj.load_state_dict(..., strict=True)
    else:
        obj.load_state_dict(...)
```

**Review by carmocca — APPROVED** [review]

File: docs/source-pytorch/fabric/api/fabric_methods.rst (hunk 5 above)
Comment (awaelchli) [reply docs/source-pytorch/fabric/api/fabric_methods.rst]: I think "must" is too strong, but we want to recommend it in general because then the strategy can access the state in its own and efficient way. BUT, if you know what you are doing, nothing prevents you from calling state_dict yourself and pass the dicts instead of objects. This is totally fine as well. 

Ok, changed the formulation

File: src/lightning_fabric/strategies/strategy.py (hunk 9 above)
Comment (awaelchli) [reply src/lightning_fabric/strategies/strategy.py]: done

**Review by Borda — APPROVED** [review]
